pkg/gatewayserver/gatewayserver.go

```diff
@@ -546,6 +548,86 @@ func (gs *GatewayServer) GetConnection(ctx context.Context, ids ttnpb.GatewayIde
 	return entry.(connectionEntry).Connection, true
 }
 
+func requireDisconnect(connected, current *ttnpb.Gateway) bool {
+	if !sameAntennaLocations(connected.Antennas, current.Antennas) {
+		// Gateway Server may update the location from status messages. If the locations aren't the same, but if the new
+		// location is a GPS location, do not disconnect the gateway. This is to avoid that updating the location from a
+		// gateway status message results in disconnecting the gateway.
+		if len(current.Antennas) > 0 && current.Antennas[0].Location != nil && current.Antennas[0].Location.Source != ttnpb.SOURCE_GPS {
+			return true
+		}
+	}
+	if connected.DownlinkPathConstraint != current.DownlinkPathConstraint ||
```

Comment:
This looks really hard to maintain (the `nil` checks got me eye balling), but I don't have a good suggestion on how to refactor it.

Reply:
Agreed. I was thinking of instantiating two new `Gateway`s and using `SetFields()` to apply the same field mask, with `connected` and `current` as sources. Then we could safely use `Equals()`. But, we're getting rid of the (generated) `Equals()` as it's a gogoproto option.